floors/rule.go

```diff
@@ -0,0 +1,404 @@
+		floorMin = rate * floorExt.FloorMin
+	}
+	return floorMin, floorCur, err
+}
+
+func updateImpExtWithFloorDetails(matchedRule string, imp *openrtb2.Imp, floorVal float64) {
+	imp.Ext, _ = jsonparser.Set(imp.Ext, []byte(`"`+matchedRule+`"`), "prebid", "floors", "floorRule")
+	imp.Ext, _ = jsonparser.Set(imp.Ext, []byte(fmt.Sprintf("%.4f", floorVal)), "prebid", "floors", "floorRuleValue")
+}
+
+func selectFloorModelGroup(modelGroups []openrtb_ext.PriceFloorModelGroup, f func(int) int) []openrtb_ext.PriceFloorModelGroup {
+	totalModelWeight := 0
+
+	for i := 0; i < len(modelGroups); i++ {
+		if modelGroups[i].ModelWeight == 0 {
+			modelGroups[i].ModelWeight = 1
+		}
+		totalModelWeight += modelGroups[i].ModelWeight
+	}
+
+	sort.SliceStable(modelGroups, func(i, j int) bool {
+		return modelGroups[i].ModelWeight < modelGroups[j].ModelWeight
+	})
+
+	winWeight := f(totalModelWeight + 1)
+	debugWeight := winWeight
+	for i, modelGroup := range modelGroups {
+		winWeight -= modelGroup.ModelWeight
+		if winWeight <= 0 {
+			modelGroups[0], modelGroups[i] = modelGroups[i], modelGroups[0]
+			modelGroups[0].DebugWeight = debugWeight
+			return modelGroups[:1]
+		}
+	}
+	return modelGroups[:1]
+}
+
+func shouldSkipFloors(ModelGroupsSkipRate, DataSkipRate, RootSkipRate int, f func(int) int) bool {
+	skipRate := 0
+
+	if ModelGroupsSkipRate > 0 {
+		skipRate = ModelGroupsSkipRate
+	} else if DataSkipRate > 0 {
+		skipRate = DataSkipRate
+	} else {
+		skipRate = RootSkipRate
+	}
+	return skipRate >= f(SKIP_RATE_MAX+1)
+}
+
+func findRule(ruleValues map[string]float64, delimiter string, desiredRuleKey []string, numFields int) (string, bool) {
+
+	ruleKeys := prepareRuleCombinations(desiredRuleKey, numFields, delimiter)
+	for i := 0; i < len(ruleKeys); i++ {
+		if _, ok := ruleValues[ruleKeys[i]]; ok {
+			return ruleKeys[i], true
+		}
+	}
+	return "", false
+}
+
+func createRuleKey(floorSchema openrtb_ext.PriceFloorSchema, request *openrtb2.BidRequest, imp openrtb2.Imp) []string {
+	var ruleKeys []string
+
+	for _, field := range floorSchema.Fields {
+		value := CATCH_ALL
+		switch field {
+		case MediaType:
+			value = getMediaType(imp)
+		case Size:
+			value = getSizeValue(imp)
+		case Domain:
+			value = getDomain(request)
+		case SiteDomain:
+			value = getSiteDomain(request)
+		case Bundle:
+			value = getBundle(request)
+		case PubDomain:
+			value = getPublisherDomain(request)
+		case Country:
+			value = getDeviceCountry(request)
+		case DeviceType:
+			value = getDeviceType(request)
+		case Channel:
+			value = extractChanelNameFromBidRequestExt(request)
+		case GptSlot:
+			value = getgptslot(imp)
+		case AdUnitCode:
+			value = getAdUnitCode(imp)
+		}
+		ruleKeys = append(ruleKeys, value)
+	}
+	return ruleKeys
+}
+
+func getDeviceType(request *openrtb2.BidRequest) string {
+	value := CATCH_ALL
+	if isMobileDevice(request.Device.UA) {
+		value = Phone
+	} else if isTabletDevice(request.Device.UA) {
+		value = Tablet
+	}
+	return value
```

Comment:
According to the requirements, `"*"` should only be returned if `request.Device.UA` does not exist so perhaps there should be an early return statement here:
```
if request.Device == nil || len(request.Device.UA) == 0 {
    return CATCH_ALL
}
```
That nil check is probably needed anyway otherwise `isMobileDevice` and `isTabletDevice` could potentially try to access a field on a `nil` value.

Reply:
added